OC-5211 Validate Public Key for Clients and Users

hosh opened the pull request:
- Added a chef_object:valid_public_key/1
- Added the valid_public_key validator to users
- Added the valid_public_key validator to clients

Related:
- https://github.com/opscode/chef_wm/pull/42
- https://github.com/opscode/chef-pedant/pull/13

hosh committed:
Added public_key validator to create user specs

hosh committed:
Added public_key validator to client specs

seth commented on src/chef_object.erl:
```diff
@@ -30,6 +30,7 @@
 -export([cert_or_key/1,
          depsolver_constraints/1,
          ejson_for_indexing/2,
+         has_public_key_header/1,
```
Is anything calling this outside this module?

seth commented on src/chef_object.erl:
```diff
@@ -709,3 +718,25 @@ key_version(<<"-----BEGIN PUBLIC KEY", _Bin/binary>>) ->
+valid_public_key(PublicKey) ->
+    case has_public_key_header(PublicKey) of
+        true ->
+            case chef_authn:extract_public_or_private_key(PublicKey) of
```
there is an `extract_public_key` export that will only return public keys. we should use that.

seth commented on src/chef_regex.erl:
```diff
@@ -135,5 +138,8 @@ regex_for(unqualified_recipe) ->
-                            <<"Malformed user name. Must only contain a-z, 0-9, _, or -">>).
+                            <<"Malformed user name. Must only contain a-z, 0-9, _, or -">>);
+
+regex_for(public_key) ->
```
I don't think this is used now, right?

hosh commented:
Ok, I'll clean this up.

seth commented on src/chef_object.erl:
```diff
@@ -623,6 +648,15 @@ value_or_null(Key, Data) ->
 
+value_or_undefined(Key, Data) ->
+  Value = ej:get(Key, Data),
+  case Value of
```
slightly more idiomatic form would be:

```
case ej:get(Key, Data) of
    null ->
        undefined;
    Value ->
        Value
end.
```

seth commented on src/chef_regex.erl:
```diff
@@ -136,4 +136,3 @@ regex_for(unqualified_recipe) ->
 regex_for(user_name) ->
    generate_regex_msg_tuple(?ANCHOR_REGEX(?USERNAME_REGEX),
                             <<"Malformed user name. Must only contain a-z, 0-9, _, or -">>).
-
```
since this file isn't really in the change set, would be nice to squash this out.

seth commented on src/chef_object.erl:
```diff
@@ -568,6 +582,17 @@ cert_or_key(Payload) ->
     end.
 
+%% Hack to get null public_key accepted as undefined
+-spec delete_null_public_key(json_object()) -> json_object().
```
I guess this makes sense as viewed as a temporary hack. If we thought this would be useful, we could have this be `delete_key_if_null(Key, Ejson)`.

hosh committed:
Client/User should accept null public key as if undefined

Some clients will send { public_key: null } and expect the API to
preserve the public key in the datastore.

Because we are using the fun_match valid_public_key, it will not
accept a null value. The kludge is to delete any null public_key before
it gets validated by ej:valid().

This should be removed when ej:valid() fun_match (or equivalent) allows
null

hosh committed:
Added public_key validator to update user specs

 - If no public_key is set, then what is in the database should not be
   changed.

 - If there is a public_key, then it must have the correct headers and
   must be convertable into a form useable by authn.

 - Adds dependency on chef_authn for key validation helper and wire
   into Makefile for dialyzer checking.

Commit:
Use chef_authn:extract_public_key as part of valid_public_key

This function will return errors for valid private keys as well as
generally invalid data.

Commit:
Add some unit test coverage for user and client JSON validation

seth commented:
@hosh I rebased on master and made some further cleanups: added needed dep on chef_authn so that unit tests run. Fixed the dialyzer spec, used the more conservative extract_public_key as noted in previous comments, updated Makefile to include chef_authn in dialyzer analysis, and added some eunit tests for chef_client and chef_user code relevant to the changes.

I'm going to go ahead and get this merged.
